# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/Lowering.cpp
# Commit: f1af63fb5986
# Full Hash: f1af63fb5986e30fa5695a26b9ed7cd506475e1d
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-09 16:10:02
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6: Force Ion to preserve its tls. r=lth
#   
#   Let's see following code and let's assume that wasm-compiler is ion:
#   
#   call foo
# ==============================================================================

diff -r facc78377809 -r f1af63fb5986 js/src/jit/Lowering.cpp
--- a/js/src/jit/Lowering.cpp	Wed Sep 09 05:43:19 2020 +0000
+++ b/js/src/jit/Lowering.cpp	Wed Sep 09 08:55:17 2020 +0000
@@ -5026,9 +5026,11 @@
 
 void LIRGenerator::visitWasmReturn(MWasmReturn* ins) {
   MDefinition* rval = ins->getOperand(0);
+  MDefinition* tlsParam = ins->getOperand(1);
 
   if (rval->type() == MIRType::Int64) {
-    add(new (alloc()) LWasmReturnI64(useInt64Fixed(rval, ReturnReg64)));
+    add(new (alloc()) LWasmReturnI64(useInt64Fixed(rval, ReturnReg64),
+                                     useFixed(tlsParam, WasmTlsReg)));
     return;
   }
 
@@ -5048,11 +5050,16 @@
     MOZ_CRASH("Unexpected wasm return type");
   }
 
+  lir->setOperand(1, useFixed(tlsParam, WasmTlsReg));
+
   add(lir);
 }
 
 void LIRGenerator::visitWasmReturnVoid(MWasmReturnVoid* ins) {
-  add(new (alloc()) LWasmReturnVoid);
+  MDefinition* tlsParam = ins->getOperand(0);
+  LWasmReturnVoid* lir = new (alloc()) LWasmReturnVoid;
+  lir->setOperand(0, useFixed(tlsParam, WasmTlsReg));
+  add(lir);
 }
 
 void LIRGenerator::visitWasmStackArg(MWasmStackArg* ins) {